### PR TITLE
Add FPMS functions and general routes to core

### PR DESCRIPTION
**Pull request by Orbitix**
I have created a number of routes that mostly mirror the tools in FPMS.

There is a large merge conflict that I don't have access to fix, and the code may need to be looked over for formatting or other mistakes.

### CODE REVIEW

**Comment by joshschmelzle** [comment]
On launch through debugging, this PR introduces a warning right away regarding with pydantic.

```
/home/wlanpi/dev/wlanpi-core/venv/lib/python3.9/site-packages/pydantic/_internal/_config.py:269: UserWarning: Valid config keys have changed in V2:
* 'allow_population_by_field_name' has been renamed to 'populate_by_name'
```

File: wlanpi_core/api/api_v1/endpoints/bluetooth_api.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+@router.get("/status", response_model=bluetooth.BluetoothStatus)
+async def btstatus():
+    """
+    Returns the bluetooth status
+    """
+
+    try:
+        status = bluetooth_service.bluetooth_status()
+        if status == False:
+            return Response(content=f"Bluetooth hardware not found", status_code=500)
```
Comment (joshschmelzle) [inline wlanpi_core/api/api_v1/endpoints/bluetooth_api.py]: 500 is the wrong HTTP code for "Bluetooth hardware not found".

File: wlanpi_core/api/api_v1/endpoints/bluetooth_api.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    # Validate action parameter
+    if action not in ["on", "off"]:
+        return Response(content="Invalid action. Use 'on' or 'off'.", status_code=400)
+
+    # Convert action to Boolean
+    state = action == "on"
+
+    if not bluetooth_service.bluetooth_present():
+        return Response(content=f"Bluetooth hardware not found", status_code=500)
```
Comment (joshschmelzle) [inline wlanpi_core/api/api_v1/endpoints/bluetooth_api.py]: Pick a different status code.

File: wlanpi_core/api/api_v1/endpoints/bluetooth_api.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    try:
+        status = bluetooth_service.bluetooth_set_power(state)
+
+        if status == False:
+            return Response(
+                content=f"Bluetooth failed to turn {action}", status_code=500
```
Comment (joshschmelzle) [inline wlanpi_core/api/api_v1/endpoints/bluetooth_api.py]: Pick a different status code.

File: wlanpi_core/api/api_v1/endpoints/utils_api.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    Runs the reachability test and returns the results
+    """
+
+    try:
+        reachability = utils_service.show_reachability()
+
+        if reachability.get("error"):
+            return Response(
+                content=json.dumps(reachability),
+                status_code=500,
```
Comment (joshschmelzle) [inline wlanpi_core/api/api_v1/endpoints/utils_api.py]: Pick a different status code.

File: wlanpi_core/api/api_v1/endpoints/utils_api.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    Runs the network speedtest and returns the results
+    """
+
+    try:
+        speedtest = utils_service.show_speedtest()
+
+        if speedtest.get("error"):
+            return Response(
+                content=json.dumps(speedtest),
+                status_code=500,
```
Comment (joshschmelzle) [inline wlanpi_core/api/api_v1/endpoints/utils_api.py]: Pick a different status code.

File: wlanpi_core/api/api_v1/endpoints/utils_api.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    Gets a list of usb interfaces and returns them.
+    """
+
+    try:
+        result = utils_service.show_usb()
+
+        if result.get("error"):
+            return Response(
+                content=json.dumps(result["error"]),
+                status_code=500,
```
Comment (joshschmelzle) [inline wlanpi_core/api/api_v1/endpoints/utils_api.py]: Pick a different status code.

File: wlanpi_core/api/api_v1/endpoints/utils_api.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    Returns the UFW information.
+    """
+
+    try:
+        result = utils_service.show_ufw()
+
+        if result.get("error"):
+            return Response(
+                content=json.dumps(result["error"]),
+                status_code=500,
```
Comment (joshschmelzle) [inline wlanpi_core/api/api_v1/endpoints/utils_api.py]: Pick a different status code.

**Comment by joshschmelzle** [comment]
A few comments:

1. Status code 500 should be used very sparingly and as a last resort exception response.
2. Error and exception handling needs to be reviewed and fixed either in this PR or a subsequent PR. Consider what you are returning back to the API consumer. Especially when there may be multiple types. UI, FPMS, Custom API, Mobile App.
```
    return JSONResponse(
        status_code=404,
        content={"message": f"<succinct message about thing API consumer wanted>"},
    )
```
3. Not excited about `speedtest-cli` making a reappearance. Is this for a demo at Prague?

**Review by joshschmelzle — COMMENTED** [review]
A few comments:

1. Status code 500 should be used very sparingly and as a last resort exception response.
2. Error and exception handling needs to be reviewed and fixed either in this PR or a subsequent PR. Consider what you are returning back to the API consumer. Especially when there may be multiple types. UI, FPMS, Custom API, Mobile App.
```
    return JSONResponse(
        status_code=404,
        content={"message": f"<succinct message about thing API consumer wanted>"},
    )
```
3. Not excited about `speedtest-cli` making a reappearance. Is this for a demo at Prague?

File: wlanpi_core/schemas/utils/utils.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    dns_server_2_resolution: Optional[str] = Field(
+        None, examples=["OK", "FAIL"], alias="DNS Server 2 Resolution"
+    )
+    dns_server_3_resolution: Optional[str] = Field(
+        None, examples=["OK", "FAIL"], alias="DNS Server 3 Resolution"
+    )
+    arping_gateway: str = Field(example="12.345ms", alias="Arping Gateway")
+
+    class Config:
+        allow_population_by_field_name = True
```
Comment (joshschmelzle) [inline wlanpi_core/schemas/utils/utils.py:22]: On launch through debugging, this PR introduces a warning right away regarding with pydantic.

```
/home/wlanpi/dev/wlanpi-core/venv/lib/python3.9/site-packages/pydantic/_internal/_config.py:269: UserWarning: Valid config keys have changed in V2:
* 'allow_population_by_field_name' has been renamed to 'populate_by_name'
```

Our pydantic version pin is 2.3.0. Please use a v2 compatible approach here.

Comment (joshschmelzle) [reply wlanpi_core/schemas/utils/utils.py:22]: https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.populate_by_name

**Comment by Orbitix** [comment]
@joshschmelzle thanks for letting me know, for the exceptions I didn't really know what the appropriate error code would be for some or them, so I can fix those. 
The speedtest route was added because it was in fpms and I was just copying from there, so I didn't know it wasn't something we wanted anymore

When do we need the error handling fixed by and should I make a new pr for it?

**Comment by joshschmelzle** [comment]
1. When the error really is on the server side, be more specific with the status code, like 501, 502, 503, 504 would probably be the ones you reach for. Maybe 511 if the endpoint is complicated and requires some sort of setup first.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

2. Please remove the code that wraps `speedtest-cli` for now. If you want to move it to a separate PR, we can discuss the motivation, support implications, performance problems with the unofficial vs official binary, EULA concerns, etc., there or on Slack. We've been down this road before.

3. If you'd like to just pick better status codes and deal with error and exception handling in another PR, that would be fine. Be consistent with what the REST API responds with. Normal returning JSON and error returning something different makes using that API difficult. Be consistent. Also make sure you represent every state accordingly. Based on quick testing last night, the responses, when working, mostly return JSON, but when there is an error, we just throw text in the response rather than JSON.

**Comment by joshschmelzle** [comment]
@Orbitix please update the versions.

**Comment by Orbitix** [comment]
> @Orbitix please update the versions.

Sorry, what do you mean by versions?
I will also address the issues once I get home later, as I can't right now

**Comment by joshschmelzle** [comment]
- Debian package version: use ~~dch -i~~ `dch -v 1.0.4` to bump the version and changelog entry. Instructions here: https://github.com/WLAN-Pi/developers/blob/main/style/DCH.md

- Python package version: edit `__version__` in `wlanpi_core/__version__.py` to match via semver without any of the extra identifiers debchange may add when you run `dch-i` like `+nmu<#>`. If you use `dch -v 1.0.4` for changelog, use `1.0.4` in `__version__.py`.

Once you do this, there should be CI that runs and checks they match.

**Comment by Orbitix** [comment]
@joshschmelzle I have fixed the issues and bumped the versions, and the error handling and status codes should be more consistent across the endpoints.

**Comment by joshschmelzle** [comment]
@Orbitix thanks. I need to fix the CI. Will do so shortly.

**Comment by joshschmelzle** [comment]
@Orbitix please fix formatting. See https://github.com/WLAN-Pi/wlanpi-core/actions/runs/11166949367/job/31042045474?pr=27.

**Review by joshschmelzle — APPROVED** [review]
Looks like we've resolved git conflicts and passed the required CI checks.

As discussed, in a subsequent PR, there needs to be better alignment on error and exception handling and consistency across what is returned across endpoints. This may require some discussion and additional documentation at an outcome in WLAN-Pi/developers. There is also opportunity to refactor some code into reusable functions that can be properly tested. Use of subprocess is a good example like we talked with Michael on Slack.

**Pull request merged**